File: server/src/main/java/cn/iinti/atom/service/base/metric/mql/Context.java
package cn.iinti.atom.service.base.metric.mql;

import cn.iinti.atom.service.base.metric.MetricEnums;
import cn.iinti.atom.service.base.metric.MetricService;
import cn.iinti.atom.service.base.metric.MetricVo;
import com.google.common.collect.Maps;
import lombok.Getter;

import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.stream.Collectors;

@Getter
public class Context {
    private final MetricEnums.MetricAccuracy metricAccuracy;

    private final MetricService metricService;

    public Context(MetricEnums.MetricAccuracy metricAccuracy, MetricService metricService) {
        this.metricAccuracy = metricAccuracy;
        this.metricService = metricService;
    }

    private final Map<String, MQLVar> variables = Maps.newHashMap();

    private final Map<String, MQLVar> exportLines = Maps.newHashMap();


    @Getter
    public static class MQLVar {
        /**
         * key is time
         */
        public TreeMap<String, List<MetricVo>> data;

        public static MQLVar newVar(TreeMap<String, List<MetricVo>> data) {
            MQLVar mqlVar = new MQLVar();
            mqlVar.data = data;
            return mqlVar;
        }

        public MQLVar copy() {
            TreeMap<String, List<MetricVo>> newData = Maps.newTreeMap();
            data.forEach((s, metricVos) -> newData.put(s, metricVos.stream().map(MetricVo::cloneMetricVo).collect(Collectors.toList())));
            return newVar(newData);
        }
    }
}

File: server/src/main/kotlin/cn/iinti/katom/base/metric/mql/Context.kt
package cn.iinti.katom.base.metric.mql

import cn.iinti.katom.base.metric.MetricEnums.MetricAccuracy
import cn.iinti.katom.base.metric.MetricService
import cn.iinti.katom.base.metric.MetricVo
import com.google.common.collect.Maps
import lombok.Getter
import java.util.*


@Getter
class Context(val metricAccuracy: MetricAccuracy, val metricService: MetricService) {
    val variables: MutableMap<String, MQLVar> = Maps.newHashMap()

    val exportLines: MutableMap<String, MQLVar> = Maps.newHashMap()


    @Getter
    class MQLVar {
        /**
         * key is time
         */
        var data: TreeMap<String, MutableList<MetricVo>>? = null

        fun copy(): MQLVar {
            val newData = Maps.newTreeMap<String, MutableList<MetricVo>>()
            data!!.forEach { (s: String?, metricVos: List<MetricVo>) ->
                newData[s] = metricVos
                    .map { obj: MetricVo -> MetricVo.cloneMetricVo(obj) }
                    .toMutableList()
            }
            return newVar(newData)
        }

        companion object {
            fun newVar(data: TreeMap<String, MutableList<MetricVo>>): MQLVar {
                val mqlVar = MQLVar()
                mqlVar.data = data
                return mqlVar
            }
        }
    }
}